app/components/SessionTimeoutHandler.tsx

```diff
@@ -0,0 +1,92 @@
+import {CiipPageComponent} from 'next-env';
+import {useRouter} from 'next/router';
+import React, {useEffect, useState} from 'react';
+import LogoutWarningModal from './LogoutWarningModal';
+
+interface Props {
+  pageComponent: CiipPageComponent;
+  modalDisplaySecondsBeforeLogout: number;
+}
+
+const SessionTimeoutHandler: React.FunctionComponent<Props> = ({
+  pageComponent,
+  modalDisplaySecondsBeforeLogout
+}) => {
+  const router = useRouter();
+
+  const [showModal, setShowModal] = useState(false);
+
+  // UNIX epoch (ms)
+  const [sessionExpiresOn, setSessionExpiresOn] = useState(Infinity);
+
+  const logoutOnSessionIdled = async () => {
+    router.push({
+      pathname: '/login-redirect',
+      query: {
+        redirectTo: router.asPath,
+        sessionIdled: true
+      }
+    });
+  };
+
+  const extendSession = async () => {
+    const response = await fetch('/extend-session');
+    if (response.ok) {
+      const timeout = Number(await response.json());
+      if (timeout > modalDisplaySecondsBeforeLogout) {
+        setShowModal(false);
+      }
+      setSessionExpiresOn(timeout * 1000 + Date.now());
+    }
+  };
+
+  useEffect(() => {
+    let timeoutId;
+
+    const checkSessionIdle = async () => {
+      if (!pageComponent.isAccessProtected) return;
+
+      const response = await fetch('/session-idle-remaining-time');
+      if (response.ok) {
+        const timeout = Number(await response.json());
+
+        setSessionExpiresOn(Date.now() + timeout * 1000);
+
+        if (timeout > modalDisplaySecondsBeforeLogout) {
+          setShowModal(false);
+          timeoutId = setTimeout(() => {
+            checkSessionIdle();
+          }, (timeout - modalDisplaySecondsBeforeLogout) * 1000);
+        } else if (timeout > 0) {
+          setShowModal(true);
+          timeoutId = setTimeout(() => {
+            checkSessionIdle();
+          }, timeout * 1000);
```

Comment:
Maybe we can add the following comment to explain the different scenarios here, as it's not obvious:

```js
// We display the modal and set a timeout to check again when the session is due to expire.
// If the user has not extended their session by then we will redirect them (by invoking logoutOnSessionIdled() below)
// If they do extend their session, the `checkSessionIdle()` call will branch into the first condition above 
// and set another timeout to check the session idle when the modal is due to be displayed.
```